File: packages/core/src/open-data.ts
export interface PublisherRef {
  readonly name: string;
  readonly url?: string;
  readonly email?: string;
}

export interface DistributionRef {
  readonly title: string;
  readonly format: string;
  readonly accessURL: string;
  readonly downloadURL?: string;
  readonly byteSize?: number;
  readonly checksum?: string;
}

const SPDX_LICENSE_URL: Record<string, string> = {
  "CC-BY-4.0": "https://creativecommons.org/licenses/by/4.0/",
  "CC-BY-3.0": "https://creativecommons.org/licenses/by/3.0/",
  "CC0-1.0": "https://creativecommons.org/publicdomain/zero/1.0/",
  "EUPL-1.2": "https://interoperable.europe.eu/collection/eupl/eupl-text-eupl-12",
};

const FORMAT_MEDIA: Record<string, string> = {
  csv: "text/csv",
  tsv: "text/tab-separated-values",
  json: "application/json",
  jsonld: "application/ld+json",
  xml: "application/xml",
  pdf: "application/pdf",
  md: "text/markdown",
  html: "text/html",
  parquet: "application/vnd.apache.parquet",
  xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
};

export function isKnownLicenseId(license: string): boolean {
  const trimmed = license.trim();
  return /^https?:\/\//i.test(trimmed) || trimmed in SPDX_LICENSE_URL;
}

export function resolveLicenseUrl(license: string): string {
  const trimmed = license.trim();
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  return SPDX_LICENSE_URL[trimmed] ?? trimmed;
}

export function resolveMediaType(format: string): string {
  const key = format.trim().toLowerCase();
  if (FORMAT_MEDIA[key]) return FORMAT_MEDIA[key]!;
  if (key.includes("/")) return key;
  return key;
}

export function parsePublisher(
  raw: unknown,
): PublisherRef | undefined {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) {
    return undefined;
  }
  const name = (raw as { name?: unknown }).name;
  if (typeof name !== "string" || name.length === 0) return undefined;
  const url = (raw as { url?: unknown }).url;
  const email = (raw as { email?: unknown }).email;
  return {
    name,
    url: typeof url === "string" && url.length > 0 ? url : undefined,
    email: typeof email === "string" && email.length > 0 ? email : undefined,
  };
}

export function parseDistributions(
  raw: unknown,
): readonly DistributionRef[] {
  if (!Array.isArray(raw)) return [];
  const out: DistributionRef[] = [];
  for (const item of raw) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) continue;
    const title = (item as { title?: unknown }).title;
    const format = (item as { format?: unknown }).format;
    const accessURL = (item as { accessURL?: unknown }).accessURL;
    if (
      typeof title !== "string" ||
      title.length === 0 ||
      typeof format !== "string" ||
      format.length === 0 ||
      typeof accessURL !== "string" ||
      accessURL.length === 0
    ) {
      continue;
    }
    const downloadURL = (item as { downloadURL?: unknown }).downloadURL;
    const byteSize = (item as { byteSize?: unknown }).byteSize;
    const checksum = (item as { checksum?: unknown }).checksum;
    out.push({
      title,
      format,
      accessURL,
      downloadURL:
        typeof downloadURL === "string" && downloadURL.length > 0
          ? downloadURL
          : undefined,
      byteSize:
        typeof byteSize === "number" && Number.isFinite(byteSize) && byteSize >= 0
          ? byteSize
          : undefined,
      checksum:
        typeof checksum === "string" && checksum.length > 0 ? checksum : undefined,
    });
  }
  return out;
}

/** Resolve distribution URL against site base (absolute URLs unchanged). */
export function resolveDistributionUrl(
  accessURL: string,
  baseUrl: string,
  pageUrl: string,
): string {
  const trimmed = accessURL.trim();
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  if (trimmed.startsWith("/")) {
    return baseUrl.length > 0 ? `${baseUrl}${trimmed}` : trimmed;
  }
  try {
    return new URL(trimmed, pageUrl).href;
  } catch {
    return trimmed;
  }
}

File: rust/sorane-astro-backend/src/open_data.rs
use serde_yaml::Value;
use std::collections::BTreeMap;

#[derive(Debug, Clone)]
pub struct PublisherRef {
    pub name: String,
    pub url: Option<String>,
    pub email: Option<String>,
}

#[derive(Debug, Clone)]
pub struct DistributionRef {
    pub title: String,
    pub format: String,
    pub access_url: String,
    pub download_url: Option<String>,
    pub byte_size: Option<u64>,
    pub checksum: Option<String>,
}

pub fn resolve_license_url(license: &str) -> String {
    let trimmed = license.trim();
    if trimmed.starts_with("http://") || trimmed.starts_with("https://") {
        return trimmed.to_string();
    }
    static SPDX: &[(&str, &str)] = &[
        ("MIT", "https://opensource.org/license/mit"),
        ("CC-BY-4.0", "https://creativecommons.org/licenses/by/4.0/"),
        ("CC-BY-3.0", "https://creativecommons.org/licenses/by/3.0/"),
        ("CC0-1.0", "https://creativecommons.org/publicdomain/zero/1.0/"),
        (
            "EUPL-1.2",
            "https://interoperable.europe.eu/collection/eupl/eupl-text-eupl-12",
        ),
    ];
    for (id, url) in SPDX {
        if trimmed == *id {
            return (*url).to_string();
        }
    }
    trimmed.to_string()
}

pub fn resolve_media_type(format: &str) -> String {
    let key = format.trim().to_lowercase();
    static FORMAT_MEDIA: &[(&str, &str)] = &[
        ("csv", "text/csv"),
        ("tsv", "text/tab-separated-values"),
        ("json", "application/json"),
        ("jsonld", "application/ld+json"),
        ("xml", "application/xml"),
        ("pdf", "application/pdf"),
        ("md", "text/markdown"),
        ("html", "text/html"),
        ("parquet", "application/vnd.apache.parquet"),
        (
            "xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        ),
    ];
    for (fmt, media) in FORMAT_MEDIA {
        if key == *fmt {
            return (*media).to_string();
        }
    }
    if key.contains('/') {
        return key;
    }
    key
}

pub fn parse_publisher(raw: &Value) -> Option<PublisherRef> {
    let map = raw.as_mapping()?;
    let name = map.get(Value::String("name".into()))?.as_str()?;
    if name.is_empty() {
        return None;
    }
    let url = map
        .get(Value::String("url".into()))
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string());
    let email = map
        .get(Value::String("email".into()))
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string());
    Some(PublisherRef {
        name: name.to_string(),
        url,
        email,
    })
}

pub fn parse_distributions(raw: Option<&Value>) -> Vec<DistributionRef> {
    let Some(Value::Sequence(items)) = raw else {
        return Vec::new();
    };
    let mut out = Vec::new();
    for item in items {
        let Some(map) = item.as_mapping() else {
            continue;
        };
        let title = map
            .get(Value::String("title".into()))
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let format = map
            .get(Value::String("format".into()))
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let access_url = map
            .get(Value::String("accessURL".into()))
            .and_then(|v| v.as_str())
            .unwrap_or("");
        if title.is_empty() || format.is_empty() || access_url.is_empty() {
            continue;
        }
        let download_url = map
            .get(Value::String("downloadURL".into()))
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string());
        let byte_size = map.get(Value::String("byteSize".into())).and_then(|v| {
            v.as_u64().or_else(|| v.as_i64().map(|n| n.max(0) as u64))
        });
        let checksum = map
            .get(Value::String("checksum".into()))
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string());
        out.push(DistributionRef {
            title: title.to_string(),
            format: format.to_string(),
            access_url: access_url.to_string(),
            download_url,
            byte_size,
            checksum,
        });
    }
    out
}

pub fn distributions_from_frontmatter(fm: &BTreeMap<String, Value>) -> Vec<DistributionRef> {
    if let Some(raw) = fm.get("distributions") {
        return parse_distributions(Some(raw));
    }
    parse_distributions(fm.get("distribution"))
}

pub fn resolve_distribution_url(access_url: &str, base_url: &str, page_url: &str) -> String {
    let trimmed = access_url.trim();
    if trimmed.starts_with("http://") || trimmed.starts_with("https://") {
        return trimmed.to_string();
    }
    if trimmed.starts_with('/') {
        return if base_url.is_empty() {
            trimmed.to_string()
        } else {
            format!("{base_url}{trimmed}")
        };
    }
    let base = if page_url.ends_with('/') {
        page_url.to_string()
    } else {
        match page_url.rsplit_once('/') {
            Some((head, _)) => format!("{head}/"),
            None => String::new(),
        }
    };
    format!("{base}{trimmed}")
}